torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -16,6 +18,139 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+class SegmentedGroup;
+class SegmentHeuristics;
+
+//! Implementation of a graph runtime with simple scheduling to support
+//! multi-kernel fusion
+class TORCH_CUDA_API FusionSegmentRuntime {
+ public:
+  //! Type notations within FusionSegmentRuntime Context
+  using HashType = size_t;
+  using SchedulerEntryPtr = std::unique_ptr<SchedulerEntry>;
+
+  explicit FusionSegmentRuntime(
+      SegmentedFusion* segmented_fusion,
+      std::unique_ptr<SegmentHeuristics>& heuristics,
+      size_t input_id);
+
+  //! FusionExecutorCache API for evicting an input id
+  void evictCache(size_t input_id) {
+    for (auto& fe : executors_) {
+      fe.evictCache(input_id);
+    }
+  }
+
+  //! FusionExecutorCache API for running the segmented fusion with given global
+  //! inputs
+  std::vector<at::Tensor> runWithInput(
+      const at::ArrayRef<IValue>& inputs,
+      size_t input_id);
+
+  //! Cache Interface: Common utility for computing hash of scheduler entires
+  static HashType getHash(SegmentHeuristics* sh);
+
+  //! Cache Interface: trivially copied and easily compared
+  //!   descriptor for FusionSegmentRuntime
+  class EntryTag {
+   public:
+    //! Computes hash upon creation
+    explicit EntryTag(SegmentHeuristics*);
+
+    //! Tag equal abstracts the heuristics equivalence
+    bool operator==(const EntryTag& other) const;
+
+    //! Returns computed hash value
+    HashType hash() const {
+      return hash_;
+    }
+
+   private:
+    HashType hash_;
+    SegmentHeuristics* heuristics_;
+  };
+
+  class EntryTagHash {
```

Comment:
can this be replaced by a simple lambda?

Reply:
Actually I was using this to supply custom cache for `map`s. It looks less confusing to me this way than a lambda but I'm more than happy to learn if there's cleaner patterns.